File: pages/20260824170700_holding_topology_passing/analysis/ruby/test/passing_schedule_test.rb
require "minitest/autorun"
require_relative "../passing_notation"
require_relative "../passing_schedule"
require_relative "../passing_occupancy"
require_relative "../passing_fixtures"

class PassingScheduleTest < Minitest::Test
  def packet(notation, beat)
    PassingAnalysis::Scheduler.schedule(notation).cycle_tosses
      .select { |event| event.beat == beat }
      .map { |event| [event.from_hand, event.to_hand, event.height, event.kind] }
  end

  def test_ultimate_routes_right_hand_passes_to_the_partner_left
    schedule = PassingAnalysis::Scheduler.schedule("<3p|3p>")
    assert_equal 6, schedule.ball_count
    assert_equal 4, schedule.hand_count
    assert_equal [[1, 2, 3, "throw"], [3, 0, 3, "throw"]], packet("<3p|3p>", 0)
    assert_equal [[0, 3, 3, "throw"], [2, 1, 3, "throw"]], packet("<3p|3p>", 1)
    assert schedule.cycle_tosses.all?(&:pass)
  end

  def test_partitioned_cascade_never_leaves_its_body
    schedule = PassingAnalysis::Scheduler.schedule("<3|3>")
    assert_equal 6, schedule.ball_count
    refute schedule.cycle_tosses.any?(&:pass)
    assert_equal [[1, 0, 3, "throw"], [3, 2, 3, "throw"]], packet("<3|3>", 0)
  end

  def test_two_count_passes_on_even_beats
    schedule = PassingAnalysis::Scheduler.schedule("<3p 3|3p 3>")
    even = schedule.cycle_tosses.select { |event| event.beat.even? }
    odd = schedule.cycle_tosses.select { |event| event.beat.odd? }
    assert even.all?(&:pass)
    refute odd.any?(&:pass)
  end

  def test_seven_club_starts_on_opposite_hands
    assert_equal [[1, 2, 4, "throw"], [2, 3, 3, "throw"]], packet("<R|L><4xp|3><3|4xp>", 0)
    assert_equal [[0, 1, 3, "throw"], [3, 0, 4, "throw"]], packet("<R|L><4xp|3><3|4xp>", 1)
  end

  def test_six_object_unit_hand_occupancy_is_half_empty_and_half_two_held
    ["<3|3>", "<3p|3p>", "<3p 3|3p 3>"].each do |notation|
      occupancy = PassingAnalysis::Occupancy.shares(
        PassingAnalysis::Scheduler.schedule(notation),
        dwell_ratio: PassingAnalysis::DWELL_RATIO
      )
      assert_equal 6, occupancy.fetch(:object_count)
      assert_equal Rational(1, 2), occupancy.fetch(:p_alpha)
      assert_equal Rational(1, 2), occupancy.fetch(:p_polymorphy)
      assert_equal 0, occupancy.fetch(:p_kappa)
      assert_equal [Rational(1, 2), 0, Rational(1, 2), 0, 0, 0, 0], occupancy.fetch(:occupancy_shares)
    end
  end

  def test_seven_club_occupancy_matches_the_two_throw_packet_law
    occupancy = PassingAnalysis::Occupancy.shares(
      PassingAnalysis::Scheduler.schedule("<R|L><4xp|3><3|4xp>"),
      dwell_ratio: PassingAnalysis::DWELL_RATIO
    )
    assert_equal 7, occupancy.fetch(:object_count)
    assert_equal Rational(1, 2), occupancy.fetch(:p_alpha)
    assert_equal Rational(1, 2), occupancy.fetch(:p_polymorphy)
    assert_equal 0, occupancy.fetch(:p_kappa)
    assert_equal [Rational(1, 2), 0, Rational(1, 2), 0, 0, 0, 0, 0], occupancy.fetch(:occupancy_shares)
  end

  def test_each_body_holds_one_or_none_in_the_six_object_unit_hand_patterns
    ["<3|3>", "<3p|3p>", "<3p 3|3p 3>"].each do |notation|
      schedule = PassingAnalysis::Scheduler.schedule(notation)
      bodies = PassingAnalysis::Occupancy.body_shares(schedule, dwell_ratio: PassingAnalysis::DWELL_RATIO)
      bodies.each do |shares|
        assert_equal Rational(1, 2), shares[0]
        assert_equal Rational(1, 2), shares[1]
        assert_equal 0, shares.drop(2).sum
      end
    end
  end

  def test_solo_cascade_writing_is_rejected
    assert_raises(ArgumentError) { PassingAnalysis::Notation.parse("3") }
  end
end

File: pages/20260824170700_holding_topology_passing/tests/passing_schedule.test.js
import assert from "node:assert/strict";
import test from "node:test";
import { passingOccupancy } from "../passing_occupancy.js";
import { schedulePassingEvents } from "../passing_schedule.js";

function packetRoutes(source, beat) {
  return schedulePassingEvents(source, true, 1).cycleTosses
    .filter((event) => event.beat === beat)
    .map((event) => [event.fromHand, event.toHand, event.height, event.kind]);
}

test("ultimate six-club passing routes 3p from each right to the partner left", () => {
  const schedule = schedulePassingEvents("<3p|3p>", true, 8);
  assert.equal(schedule.ballCount, 6);
  assert.equal(schedule.bodyCount, 2);
  assert.equal(schedule.handCount, 4);
  assert.equal(schedule.period, 1);
  assert.equal(schedule.cycleLength % 2, 0);
  assert.deepEqual(packetRoutes("<3p|3p>", 0), [
    [1, 2, 3, "throw"],
    [3, 0, 3, "throw"],
  ]);
  assert.deepEqual(packetRoutes("<3p|3p>", 1), [
    [0, 3, 3, "throw"],
    [2, 1, 3, "throw"],
  ]);
  assert.equal(schedule.cycleTosses.every((event) => event.pass), true);
});

test("partitioned double cascade never leaves its own body", () => {
  const schedule = schedulePassingEvents("<3|3>", true, 8);
  assert.equal(schedule.ballCount, 6);
  assert.equal(schedule.cycleTosses.every((event) => event.pass), false);
  assert.equal(schedule.cycleTosses.every((event) => event.fromBody === event.toBody), true);
  assert.deepEqual(packetRoutes("<3|3>", 0), [
    [1, 0, 3, "throw"],
    [3, 2, 3, "throw"],
  ]);
  assert.deepEqual(packetRoutes("<3|3>", 1), [
    [0, 1, 3, "throw"],
    [2, 3, 3, "throw"],
  ]);
});

test("two-count six keeps a pass on even beats and a self on odd beats", () => {
  const schedule = schedulePassingEvents("<3p 3|3p 3>", true, 8);
  assert.equal(schedule.ballCount, 6);
  const byBeat = Map.groupBy(schedule.cycleTosses, (event) => event.beat);
  assert.equal(byBeat.get(0).every((event) => event.pass), true);
  assert.equal(byBeat.get(1).every((event) => event.pass), false);
});

test("seven-club two-count starts on opposite hands", () => {
  const schedule = schedulePassingEvents("<R|L><4xp|3><3|4xp>", true, 16);
  assert.equal(schedule.ballCount, 7);
  assert.deepEqual(packetRoutes("<R|L><4xp|3><3|4xp>", 0), [
    [1, 2, 4, "throw"],
    [2, 3, 3, "throw"],
  ]);
  assert.deepEqual(packetRoutes("<R|L><4xp|3><3|4xp>", 1), [
    [0, 1, 3, "throw"],
    [3, 0, 4, "throw"],
  ]);
});

test("group occupancy of six-object unit-hand patterns is half empty and half two-held", () => {
  ["<3|3>", "<3p|3p>", "<3p 3|3p 3>"].forEach((notation) => {
    const occupancy = passingOccupancy(schedulePassingEvents(notation, true, 1), 0.25);
    assert.equal(occupancy.objectCount, 6);
    assert.equal(occupancy.pAlpha, 0.5);
    assert.equal(occupancy.pPolymorphy, 0.5);
    assert.equal(occupancy.pKappa, 0);
    assert.deepEqual(occupancy.occupancyShares, [0.5, 0, 0.5, 0, 0, 0, 0]);
  });
});

test("seven-club occupancy matches the two-throw packet law", () => {
  const occupancy = passingOccupancy(
    schedulePassingEvents("<R|L><4xp|3><3|4xp>", true, 1),
    0.25,
  );
  assert.equal(occupancy.objectCount, 7);
  assert.equal(occupancy.pAlpha, 0.5);
  assert.equal(occupancy.pPolymorphy, 0.5);
  assert.equal(occupancy.pKappa, 0);
  assert.deepEqual(occupancy.occupancyShares, [0.5, 0, 0.5, 0, 0, 0, 0, 0]);
});
